### ISSUE TITLE
Security hardening and bigint buffer overflow

### ISSUE THREAD

**Issue by ebfull** [zcash/zcash#2788]
There is a buffer overflow in bigint which doesn't crop up in production but does get caught when our tests have security hardening enabled. The patch for this got lost in #2252 but was revived in #2757.

We need to:

1. Enable security hardening for tests and show that tests fail with zkbot `try`.
2. Apply the patch and merge with zkbot `r+`.

**Comment by syd0** [comment]
Hi @ebfull - sorry if i'm misunderstanding what you mean, I noticed this because it was failing in the buildbot, eg https://ci.z.cash/#/builders/12/builds/133

Is this not the same as failing zkbot `try`? I'm still just learning about zcash's CI system, let me know what I'm missing. Thanks!

**Comment by ebfull** [comment]
Good question. Is the libsnark test failure not making the CI go red? @str4d

**Comment by syd0** [comment]
Would you accept making that test go green as good enough to unblock #2757 ?

**Comment by str4d** [comment]
The libsnark tests are not currently checked for hardening, and adding that in a PR could not be merged, because the tests would be red :stuck_out_tongue:

What we *should* do is make a PR that triggers the failure, run a try to show it, and then apply on top the commit that fixes it.

**Comment by str4d** [comment]
@ebfull wrote:

> Is the libsnark test failure not making the CI go red?

Oh, I think I interpreted this differently than it was intended. The reason CI itself is passing when there is a libsnark test failure, is because the test failure is on Debian 9, which is currently unsupported (so we run tests, but don't require them to pass).

Separately, I have tried enabling the hardening checks for libsnark's test binary locally, and I can't change the pass/fail stats with or without the above patch. I'll dig more closely into #2252 to see what was going on there.

**Comment by syd0** [comment]
@str4d - re: enabling hardening checks, are you talking about enabling -D_FORTIFY_SOURCE=2 or are you referring to something else?

As far as I know, I don't think FORTIFY_SOURCE is capable of detecting buffer overflow errors such as this one. I believe FORTIFY_SOURCE only helps when the overflow is caused by a few functions such as memcpy, sprintf, etc.

**Comment by str4d** [comment]
@syd0 I'm referring to the fact that in #2252, the libsnark tests were completely moved into our `zcash-gtest` binary, which has hardening applied.

I have checked out #2252 locally, reversed the patch from #2757 to undo the fix, and the built with `./zcutil/build.sh --disable-proton`. Here is the test output:

```
$ ./src/zcash-gtest --gtest_filter=libsnark.*
Note: Google Test filter = libsnark.*
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from libsnark
[ RUN      ] libsnark.bigint
*** stack smashing detected ***: ./src/zcash-gtest terminated
Aborted (core dumped)
```

If we want to close this issue per the OP, we need to figure out how to apply the same hardening to `src/snark/src/gtests` as is applied to `src/zcash-gtest`.

**Comment by syd0** [comment]
@str4d - there are a couple of things going on here:
1 - we need the `-fstack-protector-all` enabled for libsnark's build
2 - this flag only works if the method isn't being inlined.

So, I was able to reproduce the crash by adding `-fstack-protector-all` to `CXXFLAGS` in `src/snark/Makefile` **and** changing `src/Makefile.am` libsnark-tests target to pass in `-O0` in `OPTFLAGS`.

**Comment by syd0** [comment]
I just checked #2252 to see how `zcash-gtest` is being built there. Although it has `-fstack-protector-all`, it is building with `-O1`, which at least in my case causes the shorten() method to get inlined and the stack check to be bypassed.

The decision to inline shorten() at `-O1` could totally be a compiler dependent thing though. I'm building with g++ 6.3, so it could be that my compiler is being more aggressive than yours if you're building with something else.

Btw I'm still not entirely clear if you have a specific 'hardening' in mind or if you're just trying to figure out what hardening was applied to `zcash-gtest` and try apply the same thing to the libsnark gtest.

**Comment by str4d** [comment]
@daira and I decided on the following strategy:

- Make a new PR that adds `-fstack-protector-all` to `src/snark/src/gtests`
- Make a branch from that enabling `-O0`, and show it fails CI
- Cherry-pick @syd0's commit onto both branches, and show that it passes CI

This enables us to recreate @ebfull's test failure, without us enabling `-O0` in master for the tests (the motivation for which is that we want them to be as similar as possible to production).
